File: javascript/maximum-xor-for-each-query.js
/**
 * @param {number[]} nums
 * @param {number} maximumBit
 * @return {number[]}
 */
var getMaximumXor = function(nums, maximumBit) {
    let xor = nums.reduce((a, b) => a ^ b, 0);
    const mask = (1 << maximumBit) - 1;
    const result = [];
    for (let i = nums.length - 1; i >= 0; i--) {
        result.push((xor & mask) ^ mask);
        xor ^= nums[i];
    }
    return result;
};

File: elixir/maximum-xor-for-each-query.ex
defmodule Solution do
  import Bitwise, only: [bxor: 2, band: 2, bsl: 2]

  @spec get_maximum_xor(nums :: [integer], maximum_bit :: integer) :: [integer]
  def get_maximum_xor(nums, maximum_bit) do
    {result, _} = recur(bsl(1, maximum_bit) - 1, nums, 0)
    Enum.reverse(result)
  end

  defp recur(_, [], xor_sum), do: {[], xor_sum}
  defp recur(mask, [n | nums], xor_sum) do
    {result, new_xor_sum} = recur(mask, nums, bxor(n, xor_sum))

    {[bxor(band(new_xor_sum, mask), mask) | result], bxor(new_xor_sum, n)}
  end
end
